pkg/kv/kvserver/rangefeed/budget.go

```diff
@@ -0,0 +1,182 @@
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// feedBudget is memory budget for RangeFeed that wraps BoundAccount
+// and provides ability to wait for downstream to release budget and
+// to send individual events that exceed budget total budget size.
+// feedBudget doesn't provide any fairness when acquiring as it is only
+// supposed to be used by a single caller.
+// When owning component is destroyed, budget could be closed, in that
+// case all budget allocation is returned immediately and no further
+// allocations are possible.
+// In the typical case processor will get allocations from budget which
+// would be in turn borrowed from underlying account. Once event is
+// processed, allocation would be returned.
+// If budget is exhausted, allocations would block until more is available
+// or until an optional timeout expires.
+// NB: if we are competing with other consumers in the budget pool, then
+// we won't see that memory is released.
+type feedBudget struct {
+	mu struct {
+		syncutil.Mutex
+		// Bound account that provides budget with resource.
+		memBudget *mon.BoundAccount
+		// Whenever we pass oversized message this would go to true to prevent
+		// other attempts to acquire more memory until the event goes through.
+		oversize bool
+		// If true, budget was released and no more allocations could take place.
+		closed bool
+	}
+	// Channel to notify that memory was returned to the budget.
+	replenishC chan interface{}
+
+	metrics *Metrics
+}
+
+func (f *feedBudget) init(budget *mon.BoundAccount, metrics *Metrics) {
+	f.mu.memBudget = budget
+	if budget != nil {
+		f.replenishC = make(chan interface{}, 1)
+		f.metrics = metrics
+	}
+}
+
+// Get allocates amount from budget.
+// If budget has less memory than requested it could either wait, if we have any
+// in flight allocations, or it could send a single message through regardless
+// of its size. In latter case, memory is not requested from underlying budget,
+// but we won't be able to do another Get until this message is returned. To
+// track such events, feedBudget uses a counter metric.
+//
+// timeout argument indicates how long we could wait for the budget to replenish
+// if there is not enough budget available. Budget is replenished by Return
+// calls.
+// timeout value of 0 means wait indefinitely.
+// Whenever we use timeout, we return its wait chan to caller, so it could be
+// used subsequently by the chained wait for other resources thus keeping total
+// timeout in check without adding to it on every op. If we didn't initiate
+// a wait or if timeout was 0 then wait channel will be nil.
+// Returned allocation has its use counter set to 1.
+func (f *feedBudget) Get(
+	ctx context.Context, amount int64, timeout time.Duration, stoppedC <-chan struct{},
+) (sharedBudgetAllocation, <-chan time.Time, error) {
```

Comment:
This business of returning a timer channel here seems odd to me. Did you consider having this function return a 1-sized channel that the caller can wait for the allocation on, which could be prepopulated with the allocation if it succeeded immediately?

Reply:
This all stems from the original sendEvent function which tries to avoid any allocations on happy path. So it would only create wait chan if it fails on non-blocking path. We also need to wait on budget channel to get allocation and then on eventC to publish event while staying within the same timeout. There was a suggestion from Yevgeniy to use timer, but it is not straightforward as we don't have a loop where we can safely reuse and restart it.

Reply:
How about using the context? That is the standard way to handle these sorts of timeouts, after all.

For example, `sendEvent()` could do a non-blocking call to `feedBudget.TryGet()`, and if it fails, set a context timeout before trying `feedBudget.Get()` which blocks. If we fall through the `default` further down we also set a context timeout. However, context timeouts are never extended, only shortened, so if we did set a context timeout before `Get()` then that timeout will still apply later -- if we didn't, the new timeout will take effect. That seems more idiomatic to me, and would have the same performance cost -- what do you think?

Reply:
I think that would work, I changed it like that. But the monstrosity only grows